### PR TITLE
K8s 1.27 Update

### PR DESCRIPTION
**Pull request by vardhaman22**
## Issue: 
 #41840 
 
## Problem

to support kubernetes 1.27 we need to bump all the k8s libraries to 1.27.
 
## Solution

 bumping all kubernetes libraries to 1.27
 

 additional library changes
 - if we update docker form 20.10.x then there is an error regarding rancher/machine while running go mod tidy
 ```
 go: finding module for package github.com/docker/docker/pkg/term
github.com/rancher/rancher/pkg/controllers/management/drivers/nodedriver imports
        github.com/rancher/machine/libmachine/drivers/rpc imports
        github.com/rancher/machine/libmachine/drivers imports
        github.com/rancher/machine/libmachine/ssh imports
        github.com/docker/docker/pkg/term: module github.com/docker/docker@latest found (v24.0.5+incompatible), but does not contain package github.com/docker/docker/pkg/term
 ```
 that's why we need to keep docker at 20.10.x and because of that following replace statements were needed in go.mod
 ```
oras.land/oras-go => oras.land/oras-go v1.2.2
github.com/containerd/containerd => github.com/containerd/containerd  v1.6.22 
 ```

- there was an error regarding github.com/containers/image/v5/sif
```
/go/pkg/mod/github.com/containers/image/v5@v5.25.0/sif/src.go:100:3: unknown field Architecture in struct literal of type v1.Image
/go/pkg/mod/github.com/containers/image/v5@v5.25.0/sif/src.go:101:3: unknown field OS in struct literal of type v1.Image

```

that's why one more replace statement was needed
```
github.com/opencontainers/image-spec => github.com/opencontainers/image-spec v1.1.0-rc2
```
 - **FormatResourceList** function was removed from upstream. so needed to add it in a utils file. tried adding in some other existing utils files but that were causing cyclic dep error so created a new file
 
 
## Testing

dev testing is done for 1.27 clusters with forked dependencies.

Screenshot from 2023-08-17 09-47-56
Screenshot from 2023-08-17 09-36-34
Screenshot from 2023-08-17 09-29-51
Screenshot from 2023-08-17 09-29-15
Screenshot from 2023-08-17 09-21-48

### Automated Testing

* Test types added/modified:
    * Unit
    * Integration (Go Framework)
    * Integration (v2prov Framework)
    * Validation (Go Framework)
    * Other - Explain: _EXPLAIN_
    * None
    * _REMOVE NOT APPLICABLE BULLET POINTS ABOVE_
* If "None" - Reason: _EXPLAIN THE REASON_
  
* If "None" - GH Issue/PR: _LINK TO GH ISSUE/PR TO ADD TESTS_

Summary: _TODO_

### Regressions Considerations

_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

### CODE REVIEW

File: go.mod
Comment (KevinJoiner) [inline go.mod:6]: Wrangler now has a tag for this commit that can be used instead of this replace https://github.com/rancher/wrangler/tree/v2.0.1

Comment (vardhaman22) [reply go.mod:6]: actually there is an issue with adding that version in go mod. while trying to do go get we i am facing the below error
```
go: github.com/rancher/wrangler@v2.0.1: invalid version: module contains a go.mod file, so module path must match major version ("github.com/rancher/wrangler/v2")
```
so i think go requires module path to be updated when we move from v1 to v2 and that's why update in wrangler will be needed. The deadline for 1.27 was near so we decided to go with the commit tag for now.
cc: @kinarashah

Comment (kinarashah) [reply go.mod:6]: @KevinJoiner Yep, my understanding was we branched wrangler to manage branches but this isn't technically a new module for us to treat it as new v2 and change usage to `rancher/wrangler/v2`. If that's not true, we could update it in another PR to follow up.

File: Dockerfile.dapper
Comment (kinarashah) [inline Dockerfile.dapper:9]: Helm would be v3.12.3, could you update in package/Dockerfile as well

File: Dockerfile.dapper3837348211
Comment (kinarashah) [inline Dockerfile.dapper3837348211:1]: looks like this got added by mistake, make sure to remove it

File: go.mod
Comment (KevinJoiner) [reply go.mod:6]: Ah, I forgot that go requires you to change the path. We went to v2.0.1 because this version of wrangler has multiple breaking changes, so I think we need to treat it as a new module, but as you suggested we can update the paths and handle that in a separate PR.

Comment (rmweir) [reply go.mod:6]: I have a PR up for this now https://github.com/rancher/wrangler/pull/317.

File: pkg/controllers/management/test/e2e/authz_test.go
Comment (snasovich) [inline pkg/controllers/management/test/e2e/authz_test.go:343]: Should we just remove it?

Also curious why it wasn't a problem in 1.25/1.26.

File: tests/v2/integration/setup/main.go
Comment (snasovich) [inline tests/v2/integration/setup/main.go:74]: Curious why these changes were needed? I've checked the PR description and not seeing it called out explicitly.

**Review by snasovich — APPROVED** [review]
Changes look sane to me, mostly new/changed types and version bumps. Left couple minor comments / questions.

Also, want to make sure @Oats87 or @jakefhyde reviews it from v2prov standpoint as I see some CAPR changes (like in `pkg/capr/planner/certificaterotation_test.go`) that are not immediately trivial.

**Review by jiaqiluo — COMMENTED** [review]
Overall LGTM. 
Will approve after Sergey's comments are addressed.

File: pkg/capr/common.go
Comment (jakefhyde) [inline pkg/capr/common.go:344]: Checks out: https://github.com/kubernetes-sigs/cluster-api/pull/7618 :+1:

File: pkg/capr/planner/certificaterotation_test.go
Comment (jakefhyde) [inline pkg/capr/planner/certificaterotation_test.go:254]: @Oats87 correct me if I'm wrong but this is not what `joined-to` is intended to be used for. This annotation is meant to specify which server was joined to, not which controlplane object the node is a part of. I'm curious as to why this was needed.

File: pkg/controllers/capr/rkecontrolplane/controlplane.go
Comment (jakefhyde) [inline pkg/controllers/capr/rkecontrolplane/controlplane.go:106]: Does CAPI handle migrations of this label automatically? i.e. was upgrade testing performed?

File: pkg/capr/planner/certificaterotation_test.go
Comment (jakefhyde) [inline pkg/capr/planner/certificaterotation_test.go:121]: I ran the tests on 2.8-head and encountered the following:
```
=== CONT  Test_rotateCertificatesPlan/test_KCM_cert_regeneration_removal_instruction_contains_RKE2
time="2023-09-12T19:14:56-04:00" level=error msg="rkecluster /: unable to find cluster spec annotation for control plane"
time="2023-09-12T19:14:56-04:00" level=error msg="Error while generating etcd snapshot extra metadata manifest for cluster "
time="2023-09-12T19:14:56-04:00" level=error msg="rkecluster /: unable to find cluster spec annotation for control plane"
time="2023-09-12T19:14:56-04:00" level=error msg="Error while generating etcd snapshot extra metadata manifest for cluster "
time="2023-09-12T19:14:56-04:00" level=error msg="rkecluster /: unable to find cluster spec annotation for control plane"
time="2023-09-12T19:14:56-04:00" level=error msg="Error while generating etcd snapshot extra metadata manifest for cluster "
time="2023-09-12T19:14:56-04:00" level=error msg="rkecluster /: unable to find cluster spec annotation for control plane"
time="2023-09-12T19:14:56-04:00" level=error msg="rkecluster /: unable to find cluster spec annotation for control plane"
time="2023-09-12T19:14:56-04:00" level=error msg="Error while generating etcd snapshot extra metadata manifest for cluster "
time="2023-09-12T19:14:56-04:00" level=error msg="Error while generating etcd snapshot extra metadata manifest for cluster "
time="2023-09-12T19:14:56-04:00" level=error msg="rkecluster /: unable to find cluster spec annotation for control plane"
time="2023-09-12T19:14:56-04:00" level=error msg="Error while generating etcd snapshot extra metadata manifest for cluster "
time="2023-09-12T19:14:56-04:00" level=error msg="rkecluster /: unable to find cluster spec annotation for control plane"
time="2023-09-12T19:14:56-04:00" level=error msg="Error while generating etcd snapshot extra metadata manifest for cluster "
```
Are these errors still present?

File: pkg/capr/planner/config.go
Comment (jakefhyde) [inline pkg/capr/planner/config.go:139]: Why was this done? The previous function `addUserConfig` is supposed to add `MachineGlobalConfig` and `MachineSelectorConfig` to `config`, and then filter out non KDM data. Were you seeing data filtered? If so it should likely be updated in KDM.

File: pkg/controllers/capr/unmanaged/controller.go
Comment (jakefhyde) [inline pkg/controllers/capr/unmanaged/controller.go:166]: Duplicate

File: pkg/controllers/provisioningv2/provisioningcluster/template.go
Comment (jakefhyde) [inline pkg/controllers/provisioningv2/provisioningcluster/template.go:387]: Duplicate

**Review by jakefhyde — CHANGES_REQUESTED** [review]

File: pkg/controllers/capr/rkecontrolplane/controlplane.go
Comment (snasovich) [reply pkg/controllers/capr/rkecontrolplane/controlplane.go:106]: @jakefhyde , it's just the `const` name that changed in new version of CAPI it looks like.
Both `capi.ClusterLabelName` from pre-update and `capi.ClusterNameLabel` have the same value `cluster.x-k8s.io/cluster-name` so there is no "migration" to be done / upgrade testing to perform.

Am I missing something subtle here?

File: pkg/capr/planner/certificaterotation_test.go
Comment (Oats87) [reply pkg/capr/planner/certificaterotation_test.go:254]: agree, am curious as to why this is necessary

File: tests/v2/integration/setup/main.go
Comment (vardhaman22) [reply tests/v2/integration/setup/main.go:74]: This is just to reduce the errors in test..... This function sometimes errors out in the first try...so added this....it is just a copy paste from https://github.com/rancher/rancher/blob/331d18b352cfd58d5e0fc4815dffad97ca30eb49/tests/framework/extensions/pipeline/setup.go#L31

File: pkg/capr/common.go
Comment (vardhaman22) [reply pkg/capr/common.go:344]: Yes changed these constants because of cluster-api bump

File: pkg/controllers/capr/rkecontrolplane/controlplane.go
Comment (vardhaman22) [reply pkg/controllers/capr/rkecontrolplane/controlplane.go:106]: This was just a constant name change in capi so I think it should be okay.

File: pkg/capr/planner/config.go
Comment (vardhaman22) [reply pkg/capr/planner/config.go:139]: actually I did not have much experience with rke2 ...just saw that this test is broken on 2.8 head also...but it is passing in CI not sure how....so what I did is I ran the test and then tried to fix the test problems.....

For this particular thing I was getting error because even though I was setting machine global config in testdata the directory value was not right...so first I changed the machine global config key value to array of string in test data which should ideally be the case and then made this change to read these values from machinGlobal config first and use defaults otherwise.

Maybe my understanding is wrong in some cases....please let me know what should be the value here ... because the existing code breaks the test

File: pkg/capr/planner/certificaterotation_test.go
Comment (rayandas) [reply pkg/capr/planner/certificaterotation_test.go:121]: These errors were in 2.8-head and are still present

File: pkg/controllers/capr/unmanaged/controller.go
Comment (rayandas) [reply pkg/controllers/capr/unmanaged/controller.go:166]: Hi, capr.ClusterNameLabel and capi.ClusterNameLabel are different. I don't think its duplicate.

File: pkg/controllers/provisioningv2/provisioningcluster/template.go
Comment (rayandas) [reply pkg/controllers/provisioningv2/provisioningcluster/template.go:387]: Hi, capr.ClusterNameLabel and capi.ClusterNameLabel are different. I don't think its duplicate.

File: pkg/capr/planner/certificaterotation_test.go
Comment (vardhaman22) [reply pkg/capr/planner/certificaterotation_test.go]: Sorry for the confusion....my mistake....just realised it should have been set to expected joined server... @rayandas  has made the change....this annotation is required for worker only tests...https://github.com/rancher/rancher/blob/331d18b352cfd58d5e0fc4815dffad97ca30eb49/pkg/capr/planner/config.go#L110

File: pkg/capr/planner/certificaterotation_test.go
Comment (snasovich) [inline pkg/capr/planner/certificaterotation_test.go:224]: If this is not necessary, let's remove this code instead of commenting out.

**Review by snasovich — DISMISSED** [review]
One minor comment about comment to address.
But at this point it's more about getting @jakefhyde @Oats87 to circle back.

Approving assuming that nit is addressed.

File: pkg/controllers/capr/rkecontrolplane/controlplane.go
Comment (jakefhyde) [reply pkg/controllers/capr/rkecontrolplane/controlplane.go:106]: Could've sworn I deleted this comment prior to leaving my review, oh well.

File: pkg/capr/planner/certificaterotation_test.go
Comment (rayandas) [reply pkg/capr/planner/certificaterotation_test.go]: @snasovich updated the PR

File: pkg/controllers/capr/unmanaged/controller.go
Comment (jakefhyde) [reply pkg/controllers/capr/unmanaged/controller.go:166]: It is a duplicate of the line directly above

Comment (rayandas) [reply pkg/controllers/capr/unmanaged/controller.go:166]: No problem. :)

**Comment by jakefhyde** [comment]
I have created an [issue](https://github.com/rancher/rancher/issues/42793) and subsequent [PR](https://github.com/rancher/rancher/pull/42794) to fix the issue you are seeing with the cert rotation test. After that is merged, can you revert your changes to `certificaterotation_test.go`?

**Comment by snasovich** [comment]
Superseded by https://github.com/rancher/rancher/pull/42808

**Pull request closed without merging**
